File: 05-failure/02-timeout-budgets-and-deadline-propagation/java/DeadlineChain.java
// Layer 5 - Topic 2: deadline propagation through a three-hop chain, in one
// JVM.
//
// Java sits between Go and Rust on this topic. There is no ambient context
// in the platform -- gRPC-Java ships its own Context/Deadline, and Deadline
// is absolute-time based, which is the shape modelled below.
// HttpClient.newBuilder().connectTimeout() plus
// HttpRequest.newBuilder().timeout() cover the transport;
// CompletableFuture.orTimeout covers composition; and JDBC's
// Statement.setQueryTimeout is the equivalent of statement_timeout -- with
// the same caveat, that it ASKS THE DRIVER to cancel, and what the server
// does about that is the server's business.
//
// Java 21's ScopedValue plus StructuredTaskScope is the first thing in the
// platform that looks like Go's context tree. Both are still preview APIs in
// 21, so this file uses a ThreadLocal to carry the deadline instead -- which
// is exactly what gRPC-Java's Context does underneath, and which works
// unchanged on virtual threads. If you have virtual threads and a JDK where
// ScopedValue is final, that is the shape to build; the mechanism this file
// demonstrates does not change.
//
// WHAT THIS DEMONSTRATES
//
//   gateway -> serviceB -> serviceC, C holds a pooled connection for a
//   controlled service time, gateway budget 500ms.
//
//     1 healthy               everything succeeds; the bug is invisible
//     2 naive                 every hop uses the same 500ms constant, so B
//                             and C never learn what is left of the budget
//     3 deadline propagated   the absolute deadline rides a ThreadLocal;
//                             B and C refuse work that cannot finish, and
//                             hand a connection straight back when the
//                             request behind it is already dead
//     4 + setQueryTimeout     the query itself is bounded, not just the
//                             CompletableFuture waiting for it
//
// WHAT TO LOOK FOR IN THE OUTPUT
//   1. `zombie/s` -- completions C finished AFTER the gateway had already
//      returned 504. One pool slot and one service time each.
//   2. `C pool in use` pinned at the pool size in row 2. That is topic 1's
//      L, spent entirely on work nobody is waiting for.
//   3. Row 3 helps and does not fix it. orTimeout completes YOUR future
//      exceptionally; it does not reach into the database. This is the same
//      finding as Python's shield and Rust's spawn_blocking, in a third
//      spelling.
//   4. Row 4 bounds the work at the resource, and the pool comes back down.
//      Watch gateway success and `C pool in use` move together -- they are
//      the same fact stated twice.
//
// The load generator is OPEN MODEL: Poisson arrivals, and it does not wait
// for a response before sending the next request.
//
// RUN
//   javac DeadlineChain.java -d /tmp/javabuild && java -cp /tmp/javabuild DeadlineChain

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Random;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.Semaphore;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicLong;

public class DeadlineChain {

    // -------------------------------------------------------------- config

    static final long GATEWAY_BUDGET_MS = 500;
    static final long SLACK_MS = 20;
    static final long HOP_OVERHEAD_MS = 5;
    static final long C_SERVICE_FAST_MS = 40;
    static final long C_SERVICE_SLOW_MS = 800;
    static final double SLOW_FRACTION = 0.25;
    static final int C_POOL_SIZE = 8;
    static final double RATE = 50.0;
    static final long DURATION_MS = 12_000;
    static final long WARMUP_MS = 3_000;   // JIT only; never printed
    static final long GAUGE_EVERY_MS = 20;

    // ------------------------------------------------------ the deadline

    /**
     * An ABSOLUTE deadline, which is what gRPC-Java's Deadline is too. The
     * absoluteness is the load-bearing part: a relative per-call timeout
     * cannot compose, because the third call in a handler has no idea what
     * the first two already spent.
     */
    record Deadline(long atNanos) {
        long remainingMs() {
            return TimeUnit.NANOSECONDS.toMillis(atNanos - System.nanoTime());
        }
        boolean expiredWithin(long marginMs) {
            return remainingMs() < marginMs;
        }
    }

    /**
     * The carrier. gRPC-Java's Context is a ThreadLocal underneath, and a
     * ThreadLocal is inherited by nothing -- so every place the deadline
     * needs to cross a thread boundary is a place somebody has to write code.
     * That is the whole ergonomic gap between this and Go's ctx parameter.
     */
    static final ThreadLocal<Deadline> DEADLINE = new ThreadLocal<>();

    static Deadline deadline() {
        return DEADLINE.get();
    }

    // ------------------------------------------------------------ metrics

    static final class Metrics {
        final AtomicLong ok = new AtomicLong();
        final AtomicLong failed = new AtomicLong();
        final AtomicLong zombie = new AtomicLong();
        final AtomicLong killed = new AtomicLong();
        final AtomicLong abandoned = new AtomicLong();
        final AtomicLong inUse = new AtomicLong();
        final List<Double> cLatency = Collections.synchronizedList(new ArrayList<>());
        final List<Double> gauge = Collections.synchronizedList(new ArrayList<>());
    }

    // --------------------------------------------------------- the pool

    /**
     * HikariCP, and the database behind it. A permit is a connection.
     * Nothing the caller does shortens a query that is already running --
     * only the query timeout does, and only when one was set.
     */
    static final class Pool {
        private final Semaphore permits;
        private final Metrics m;

        Pool(int size, Metrics m) {
            this.permits = new Semaphore(size, true);
            this.m = m;
        }

        boolean query(long durationMs, Deadline dl, boolean useQueryTimeout)
                throws InterruptedException {
            permits.acquire();
            try {
                // Checked out. If the request that queued for this connection
                // died while it was queueing, give the connection straight
                // back rather than spend a whole service time on a corpse.
                // Under overload this is where the recovered capacity is.
                if (dl != null && dl.expiredWithin(SLACK_MS)) {
                    m.abandoned.incrementAndGet();
                    return false;
                }

                m.inUse.incrementAndGet();
                try {
                    if (dl != null && useQueryTimeout) {
                        // Statement.setQueryTimeout(seconds), derived from the
                        // SAME number as the application budget. Two
                        // independently chosen timeouts is how you end up
                        // shedding load in the app while the database stays
                        // pinned -- the errors AND the load.
                        long budget = Math.max(0, dl.remainingMs() - SLACK_MS);
                        if (budget < durationMs) {
                            Thread.sleep(budget);
                            m.killed.incrementAndGet();
                            return false;
                        }
                    }
                    Thread.sleep(durationMs);
                    return true;
                } finally {
                    m.inUse.decrementAndGet();
                }
            } finally {
                permits.release();
            }
        }
    }

    static final class Expired extends RuntimeException {
        Expired(String s) {
            super(s, null, false, false);
        }
    }

    // ---------------------------------------------------------- the hops

    static void serviceC(Pool pool, Metrics m, boolean slow, long gatewayDeadlineNanos,
                         boolean useQueryTimeout) throws InterruptedException {
        Deadline dl = deadline();
        if (dl != null && dl.expiredWithin(SLACK_MS)) {
            // Refuse to START work that cannot finish. A request rejected
            // here costs no pool slot, no queue position, nothing at all.
            throw new Expired("no budget left at C");
        }
        Thread.sleep(HOP_OVERHEAD_MS);

        long duration = slow ? C_SERVICE_SLOW_MS : C_SERVICE_FAST_MS;
        long started = System.nanoTime();

        // The query runs on its own virtual thread -- a database session,
        // not a coroutine of yours. Completing our CompletableFuture
        // exceptionally below does not touch it.
        CompletableFuture<Boolean> query = new CompletableFuture<>();
        Thread.ofVirtual().start(() -> {
            boolean completed;
            try {
                completed = pool.query(duration, dl, useQueryTimeout);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                completed = false;
            }
            long finished = System.nanoTime();
            m.cLatency.add((finished - started) / 1e6);
            if (completed && finished > gatewayDeadlineNanos) {
                m.zombie.incrementAndGet();
            }
            query.complete(completed);
        });

        long localBudget = dl != null
                ? Math.max(0, dl.remainingMs())
                : GATEWAY_BUDGET_MS;   // the naive constant, copied down the chain

        try {
            // orTimeout completes THIS stage exceptionally. The virtual
            // thread above never finds out. That sentence is the topic.
            query.orTimeout(localBudget, TimeUnit.MILLISECONDS).join();
        } catch (Exception e) {
            throw new Expired("timed out waiting on C");
        }
    }

    static void serviceB(Pool pool, Metrics m, boolean slow, long gatewayDeadlineNanos,
                         boolean useQueryTimeout) throws InterruptedException {
        Deadline dl = deadline();
        if (dl != null && dl.expiredWithin(SLACK_MS)) {
            throw new Expired("no budget left at B");
        }
        Thread.sleep(HOP_OVERHEAD_MS);

        if (dl != null) {
            // budget_out = budget_in - elapsed_here - slack. With an absolute
            // deadline the subtraction is the only arithmetic there is, and
            // it can only ever tighten.
            DEADLINE.set(new Deadline(dl.atNanos() - TimeUnit.MILLISECONDS.toNanos(SLACK_MS)));
        }
        serviceC(pool, m, slow, gatewayDeadlineNanos, useQueryTimeout);
    }

    static void gateway(Pool pool, Metrics m, boolean slow, boolean propagate,
                        boolean useQueryTimeout) {
        long gatewayDeadlineNanos =
                System.nanoTime() + TimeUnit.MILLISECONDS.toNanos(GATEWAY_BUDGET_MS);
        if (propagate) {
            DEADLINE.set(new Deadline(gatewayDeadlineNanos));
        } else {
            DEADLINE.remove();
        }
        try {
            serviceB(pool, m, slow, gatewayDeadlineNanos, useQueryTimeout);
            // The gateway's own budget, enforced locally as well as passed
            // on: a caller that only trusts the callee to be timely has no
            // timeout at all.
            if (System.nanoTime() <= gatewayDeadlineNanos) {
                m.ok.incrementAndGet();
            } else {
                m.failed.incrementAndGet();
            }
        } catch (Exception e) {
            m.failed.incrementAndGet();
        } finally {
            DEADLINE.remove();
        }
    }

    // -------------------------------------------------------- the driver

    static Metrics runVariant(double slowFraction, boolean propagate, boolean useQueryTimeout)
            throws Exception {
        return runVariant(slowFraction, propagate, useQueryTimeout, DURATION_MS);
    }

    static Metrics runVariant(double slowFraction, boolean propagate, boolean useQueryTimeout,
                              long durationMs)
            throws InterruptedException {
        Metrics m = new Metrics();
        Pool pool = new Pool(C_POOL_SIZE, m);
        // Identical arrivals and an identical set of slow requests in every
        // variant, so what differs between rows is policy and only policy.
        Random rng = new Random(20250502);

        Thread sampler = Thread.ofPlatform().daemon().start(() -> {
            try {
                while (!Thread.currentThread().isInterrupted()) {
                    Thread.sleep(GAUGE_EVERY_MS);
                    m.gauge.add((double) m.inUse.get());
                }
            } catch (InterruptedException ignored) {
            }
        });

        long begin = System.nanoTime();
        long end = begin + TimeUnit.MILLISECONDS.toNanos(durationMs);
        long at = begin;
        List<Thread> requests = new ArrayList<>();
        while (true) {
            at += (long) (-Math.log(1 - rng.nextDouble()) / RATE * 1e9);
            if (at > end) break;
            long wait = at - System.nanoTime();
            if (wait > 0) Thread.sleep(wait / 1_000_000L, (int) (wait % 1_000_000L));
            boolean slow = rng.nextDouble() < slowFraction;
            // One virtual thread per request. On platform threads this
            // program would be a study in thread exhaustion instead; see the
            // C++ version's `threads peak` column for what that looks like.
            requests.add(Thread.ofVirtual().start(
                    () -> gateway(pool, m, slow, propagate, useQueryTimeout)));
        }
        for (Thread t : requests) t.join();
        // Drain. Zombies are by definition still running after everyone gave
        // up, so a report taken at the end of the load would undercount them.
        Thread.sleep(C_SERVICE_SLOW_MS + 300);
        sampler.interrupt();
        return m;
    }

    // ------------------------------------------------------- reporting

    static final String HEADER =
            "variant                      gw success  zombie/s  C pool in use  C p99 ms  killed/s  gaveback/s";

    static void printRow(String label, Metrics m) {
        double seconds = DURATION_MS / 1000.0;
        long total = m.ok.get() + m.failed.get();
        double success = total > 0 ? 100.0 * m.ok.get() / total : 0;
        List<Double> lat = new ArrayList<>(m.cLatency);
        Collections.sort(lat);
        System.out.printf("%-28s %9.1f%% %9.1f %13s %9.0f %9.1f %11.1f%n",
                label, success, m.zombie.get() / seconds,
                String.format("%.1f/%d", mean(m.gauge), C_POOL_SIZE),
                percentile(lat, 99), m.killed.get() / seconds, m.abandoned.get() / seconds);
    }

    static double percentile(List<Double> sorted, double p) {
        if (sorted.isEmpty()) return 0;
        int k = (int) Math.round(p / 100 * (sorted.size() - 1));
        return sorted.get(Math.min(k, sorted.size() - 1));
    }

    static double mean(List<Double> v) {
        List<Double> copy = new ArrayList<>(v);
        if (copy.isEmpty()) return 0;
        double s = 0;
        for (double x : copy) s += x;
        return s / copy.size();
    }

    public static void main(String[] args) throws Exception {
        double fastDemand = RATE * (1 - SLOW_FRACTION) * C_SERVICE_FAST_MS / 1000.0;
        double slowDemand = RATE * SLOW_FRACTION * C_SERVICE_SLOW_MS / 1000.0;

        System.out.printf("Deadline propagation through gateway -> serviceB -> serviceC (%s).%n",
                Runtime.version());
        System.out.printf("Gateway budget %dms, slack %dms per hop, C pool %d, offered %.0f rps for %ds.%n",
                GATEWAY_BUDGET_MS, SLACK_MS, C_POOL_SIZE, RATE, DURATION_MS / 1000);
        System.out.printf("When C is unwell, %.0f%% of queries take %dms and the rest take %dms.%n",
                SLOW_FRACTION * 100, C_SERVICE_SLOW_MS, C_SERVICE_FAST_MS);
        System.out.printf("Demand on the pool is then %.1f + %.1f = %.1f connection-seconds per second%n",
                slowDemand, fastDemand, slowDemand + fastDemand);
        System.out.printf("against %d available, i.e. rho = %.2f. None of the slow queries can beat the budget.%n%n",
                C_POOL_SIZE, (slowDemand + fastDemand) / C_POOL_SIZE);
        System.out.println(HEADER);
        System.out.println("-".repeat(HEADER.length()));

        // Discarded. The first variant to run pays for class loading and for
        // the interpreter running every hop until C2 catches up, and on a
        // 500ms budget that alone fails ~7% of requests and drags C's p99 to
        // ~450ms. Left in, it would make the healthy BASELINE look like a
        // symptom of the thing this program is about. Warm it, then measure.
        runVariant(0.0, false, false, WARMUP_MS);

        printRow("1 healthy", runVariant(0.0, false, false));
        printRow("2 naive", runVariant(SLOW_FRACTION, false, false));
        printRow("3 deadline propagated", runVariant(SLOW_FRACTION, true, false));
        printRow("4 + setQueryTimeout", runVariant(SLOW_FRACTION, true, true));

        System.out.println();
        System.out.println("Rows 2 and 3: an absolute deadline on a ThreadLocal, subtracted once");
        System.out.println("per hop, is the whole discipline. B and C stop queueing work whose");
        System.out.println("caller has already gone, and hand connections back the moment they");
        System.out.println("find one checked out for a dead request ('gaveback/s').");
        System.out.println();
        System.out.println("Rows 3 and 4 are the half the platform will not do for you.");
        System.out.println("CompletableFuture.orTimeout completes YOUR stage exceptionally. The");
        System.out.println("virtual thread holding the connection never hears about it, and");
        System.out.println("neither does the database. setQueryTimeout is the piece that bounds");
        System.out.println("the work rather than the waiting -- and even it only asks the driver");
        System.out.println("to cancel. What the server does about that request is the server's");
        System.out.println("business, which is why Postgres has statement_timeout as well.");
    }
}

File: 05-failure/02-timeout-budgets-and-deadline-propagation/golang/deadline_chain.go
// Layer 5 - Topic 2: deadline propagation through a three-hop chain, in one
// Go process.
//
// Go is genuinely ahead of every other runtime in this layer on this topic,
// and it is worth reading the source of `context` -- it is short -- as the
// reference implementation of the idea. context.WithTimeout produces a value
// that database/sql, net/http and every well-behaved library obey;
// r.Context() is ALREADY cancelled when the client disconnects; cancellation
// propagates down the tree without anyone opting in.
//
// Which is why Go's failure mode is the inverse of everyone else's. Nobody
// forgets to pass a context in Go; the compiler and every linter shout. What
// people do instead is write context.Background() somewhere in the middle of
// a handler -- to "start something that outlives the request", or because a
// helper's signature did not take one -- and silently detach an entire
// subtree from the deadline. Variant 2 below is exactly that one line.
//
// WHAT THIS DEMONSTRATES
//
//	gateway -> serviceB -> serviceC, where C holds a pooled connection for a
//	controlled service time. The gateway's budget is 500ms.
//
//	 1. healthy               everything succeeds; the bug is invisible
//	 2. detached at B         serviceB calls context.Background(): one line,
//	                          no error, no warning, no deadline at all below
//	                          it, and C never learns that the gateway gave up
//	 3. context threaded      the deadline reaches C, which refuses work it
//	                          cannot finish and hands a connection back the
//	                          moment it finds the request behind it is dead
//	 4. + QueryContext        the query itself honours the context, so the
//	                          connection is released when the caller gives
//	                          up rather than when the query finishes
//
// WHAT TO LOOK FOR IN THE OUTPUT
//  1. Rows 1 and 2 are the same code with one line different, and the
//     difference is a service that works and a service that does not.
//  2. `zombie/s` -- completions C finished after the gateway had already
//     returned 504. Each one is a pool slot and a service time spent
//     producing a response nobody will read.
//  3. `C pool in use` pinned at the pool size in row 2. That is topic 1's
//     L, consumed entirely by dead work.
//  4. Row 3 versus row 4: threading the context is not the same as the
//     driver honouring it. A pool slot is held until the QUERY returns,
//     not until your goroutine gives up waiting for it.
//
// The load generator is OPEN MODEL: Poisson arrivals, and it does not wait
// for a response before sending the next request.
//
// RUN
//
//	go run deadline_chain.go
package main

import (
	"context"
	"fmt"
	"math"
	"math/rand"
	"sort"
	"sync"
	"sync/atomic"
	"time"
)

// ------------------------------------------------------------------ config

const (
	gatewayBudget = 500 * time.Millisecond
	slack         = 20 * time.Millisecond
	hopOverhead   = 5 * time.Millisecond
	cServiceFast  = 40 * time.Millisecond
	cServiceSlow  = 800 * time.Millisecond
	slowFraction  = 0.25
	cPoolSize     = 8
	rate          = 50.0
	duration      = 12 * time.Second
	gaugeEvery    = 20 * time.Millisecond
)

// ----------------------------------------------------------------- metrics

type metrics struct {
	mu        sync.Mutex
	ok        int64
	failed    int64
	zombie    int64
	killed    int64
	abandoned int64
	cLatency  []float64
	gauge     []float64
}

func (m *metrics) observeC(latency float64, zombie bool) {
	m.mu.Lock()
	m.cLatency = append(m.cLatency, latency)
	m.mu.Unlock()
	if zombie {
		atomic.AddInt64(&m.zombie, 1)
	}
}

// -------------------------------------------------------------- the pool

// pool is what database/sql's MaxOpenConns is underneath, plus a database
// server on the other end of it. The server does not care about your
// goroutines: query holds a connection for its whole duration unless the
// driver was given a context it actually honours.
type pool struct {
	tokens chan struct{}
	inUse  atomic.Int64
	m      *metrics
}

func newPool(size int, m *metrics) *pool {
	p := &pool{tokens: make(chan struct{}, size), m: m}
	for i := 0; i < size; i++ {
		p.tokens <- struct{}{}
	}
	return p
}

// query acquires a connection and runs for d.
//
// honourContext models the difference between db.Query(...) and
// db.QueryContext(ctx, ...). With it false the statement runs to completion
// no matter what the caller does -- which is also what happens to a real
// QueryContext once the rows are already coming back, and what happens to
// anything the driver has handed to the server.
func (p *pool) query(ctx context.Context, d time.Duration, honourContext bool) bool {
	// Waiting for a connection is itself cancellable, which is the part
	// database/sql gets right and most hand-rolled pools do not.
	select {
	case <-p.tokens:
	case <-ctx.Done():
		return false
	}
	defer func() { p.tokens <- struct{}{} }()

	// Checked out. If the request that queued for this connection died while
	// it was queueing, give the connection straight back rather than spend a
	// service time on a corpse. Under overload this is where most of the
	// recovered capacity comes from.
	if dl, ok := ctx.Deadline(); ok && time.Until(dl) < slack {
		atomic.AddInt64(&p.m.abandoned, 1)
		return false
	}

	p.inUse.Add(1)
	defer p.inUse.Add(-1)

	if !honourContext {
		time.Sleep(d)
		return true
	}
	select {
	case <-time.After(d):
		return true
	case <-ctx.Done():
		atomic.AddInt64(&p.m.killed, 1)
		return false
	}
}

// --------------------------------------------------------------- the hops

func serviceC(ctx context.Context, p *pool, m *metrics, slow bool,
	gatewayDeadline time.Time, honourContext bool) error {

	if dl, ok := ctx.Deadline(); ok && time.Until(dl) < slack {
		// Refuse to START work that cannot finish. A request rejected here
		// costs no pool slot, no queue position, nothing at all.
		return context.DeadlineExceeded
	}
	time.Sleep(hopOverhead)

	d := cServiceFast
	if slow {
		d = cServiceSlow
	}
	started := time.Now()

	// The query runs in its own goroutine so that this hop can stop waiting
	// without the query stopping. That is not an artefact of the test: it is
	// what a database server does, and what any driver call already in
	// flight does, in every language in this layer.
	done := make(chan bool, 1)
	go func() {
		completed := p.query(ctx, d, honourContext)
		finished := time.Now()
		m.observeC(float64(finished.Sub(started).Milliseconds()),
			completed && finished.After(gatewayDeadline))
		done <- completed
	}()

	select {
	case <-done:
		return nil
	case <-ctx.Done():
		return ctx.Err()
	}
}

func serviceB(ctx context.Context, p *pool, m *metrics, slow bool,
	gatewayDeadline time.Time, detach bool, honourContext bool) error {

	if dl, ok := ctx.Deadline(); ok && time.Until(dl) < slack {
		return context.DeadlineExceeded
	}
	time.Sleep(hopOverhead)

	outCtx := ctx
	var cancel context.CancelFunc
	if detach {
		// ==================== THE ONE LINE ====================
		// It compiles, it passes review, `go vet` is silent, and the
		// linter that complains about a missing context is satisfied,
		// because there IS a context. It just is not the caller's -- and
		// context.Background() carries no deadline, exactly like the zero
		// value of http.Client, so everything below this point now has
		// infinite patience for a request the gateway already abandoned.
		outCtx, cancel = context.WithCancel(context.Background())
		// ======================================================
	} else {
		// budget_out = budget_in - elapsed_here - slack. With a context you
		// mostly do not even write this: passing ctx down already carries
		// the absolute deadline, and WithTimeout here only ever TIGHTENS it.
		outCtx, cancel = context.WithTimeout(ctx, time.Until(gatewayDeadline)-slack)
	}
	defer cancel()

	return serviceC(outCtx, p, m, slow, gatewayDeadline, honourContext)
}

func gateway(p *pool, m *metrics, slow, detach, honourContext bool) {
	// In a real server this is r.Context() with a timeout on top, and it is
	// already cancelled if the client hangs up. Nobody has to remember.
	ctx, cancel := context.WithTimeout(context.Background(), gatewayBudget)
	defer cancel()
	deadline, _ := ctx.Deadline()

	// The gateway enforces its own budget locally as well as passing it on,
	// because a caller that only trusts the callee to be timely has no
	// timeout at all. This select is also the moment the gateway stops
	// waiting -- and, in variant 2, the moment nothing downstream notices.
	errc := make(chan error, 1)
	go func() { errc <- serviceB(ctx, p, m, slow, deadline, detach, honourContext) }()
	select {
	case err := <-errc:
		if err != nil {
			atomic.AddInt64(&m.failed, 1)
		} else {
			atomic.AddInt64(&m.ok, 1)
		}
	case <-ctx.Done():
		atomic.AddInt64(&m.failed, 1)
	}
}

// -------------------------------------------------------------- the driver

func runVariant(slowFrac float64, detach, honourContext bool) *metrics {
	m := &metrics{}
	p := newPool(cPoolSize, m)
	// Identical arrivals and an identical set of slow requests in every
	// variant, so what differs between the rows is policy and only policy.
	rng := rand.New(rand.NewSource(20250502))

	stop := make(chan struct{})
	var wg sync.WaitGroup
	wg.Add(1)
	go func() {
		defer wg.Done()
		t := time.NewTicker(gaugeEvery)
		defer t.Stop()
		for {
			select {
			case <-stop:
				return
			case <-t.C:
				m.mu.Lock()
				m.gauge = append(m.gauge, float64(p.inUse.Load()))
				m.mu.Unlock()
			}
		}
	}()

	begin := time.Now()
	end := begin.Add(duration)
	at := begin
	var reqs sync.WaitGroup
	for {
		at = at.Add(time.Duration(rng.ExpFloat64() / rate * float64(time.Second)))
		if at.After(end) {
			break
		}
		if d := time.Until(at); d > 0 {
			time.Sleep(d)
		}
		slow := rng.Float64() < slowFrac
		reqs.Add(1)
		go func() {
			defer reqs.Done()
			gateway(p, m, slow, detach, honourContext)
		}()
	}
	reqs.Wait()
	// Drain. Zombies are by definition still running after everyone gave up,
	// so a report taken at the end of the load would undercount them.
	time.Sleep(cServiceSlow + 300*time.Millisecond)
	close(stop)
	wg.Wait()
	return m
}

// -------------------------------------------------------------- reporting

const header = "variant                      gw success  zombie/s  C pool in use  C p99 ms  killed/s  gaveback/s"

func printRow(label string, m *metrics) {
	seconds := duration.Seconds()
	total := m.ok + m.failed
	success := 0.0
	if total > 0 {
		success = 100 * float64(m.ok) / float64(total)
	}
	m.mu.Lock()
	lat := append([]float64(nil), m.cLatency...)
	gauge := append([]float64(nil), m.gauge...)
	m.mu.Unlock()
	sort.Float64s(lat)

	fmt.Printf("%-28s %9.1f%% %9.1f %13s %9.0f %9.1f %11.1f\n",
		label, success, float64(m.zombie)/seconds,
		fmt.Sprintf("%.1f/%d", mean(gauge), cPoolSize),
		percentile(lat, 99), float64(m.killed)/seconds,
		float64(m.abandoned)/seconds)
}

func percentile(sorted []float64, p float64) float64 {
	if len(sorted) == 0 {
		return 0
	}
	k := int(math.Round(p / 100 * float64(len(sorted)-1)))
	if k >= len(sorted) {
		k = len(sorted) - 1
	}
	return sorted[k]
}

func mean(v []float64) float64 {
	if len(v) == 0 {
		return 0
	}
	s := 0.0
	for _, x := range v {
		s += x
	}
	return s / float64(len(v))
}

func main() {
	fastDemand := rate * (1 - slowFraction) * cServiceFast.Seconds()
	slowDemand := rate * slowFraction * cServiceSlow.Seconds()

	fmt.Println("Deadline propagation through gateway -> serviceB -> serviceC, in Go.")
	fmt.Printf("Gateway budget %v, slack %v per hop, C pool %d, offered %.0f rps for %v.\n",
		gatewayBudget, slack, cPoolSize, rate, duration)
	fmt.Printf("When C is unwell, %.0f%% of queries take %v and the rest take %v.\n",
		slowFraction*100, cServiceSlow, cServiceFast)
	fmt.Printf("Demand on the pool is then %.1f + %.1f = %.1f connection-seconds per second\n",
		slowDemand, fastDemand, slowDemand+fastDemand)
	fmt.Printf("against %d available, i.e. rho = %.2f. None of the slow queries can beat the budget.\n\n",
		cPoolSize, (slowDemand+fastDemand)/cPoolSize)
	fmt.Println(header)
	fmt.Println(dashes(len(header)))

	printRow("1 healthy", runVariant(0, false, false))
	printRow("2 detached at B", runVariant(slowFraction, true, false))
	printRow("3 context threaded", runVariant(slowFraction, false, false))
	printRow("4 + QueryContext", runVariant(slowFraction, false, true))

	fmt.Println()
	fmt.Println("Rows 2 and 3 differ by exactly one line -- context.Background() instead")
	fmt.Println("of the caller's ctx -- and that line is the difference between a chain")
	fmt.Println("that sheds dead work and one that works hardest precisely when nothing")
	fmt.Println("it produces can be delivered. There is no error, no warning, no vet")
	fmt.Println("diagnostic. Grep your codebase for context.Background() below a handler")
	fmt.Println("boundary; that grep is worth more than most of this program.")
	fmt.Println()
	fmt.Println("Rows 3 and 4 are the part context alone does not buy you. Threading a")
	fmt.Println("deadline tells your code when to stop waiting; a connection is held")
	fmt.Println("until the QUERY returns. db.QueryContext closes that gap -- and even")
	fmt.Println("then only up to the point where the statement is already executing on")
	fmt.Println("the server, which is why Postgres has statement_timeout at all.")
}

func dashes(n int) string {
	b := make([]byte, n)
	for i := range b {
		b[i] = '-'
	}
	return string(b)
}
